<a href="https://colab.research.google.com/github/Tchalz/delete_now/blob/master/ASSIGNMENT_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## SECTION 1: DATA EXPLORATION

###QUESTION 1 ANSWERS

In [3]:
import pandas as pd
import numpy as np
df = pd.read_csv(r'/content/housing_dataset.csv')
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15,5612,1283.0,1015.0,472,1.4936,66900
1,-114.47,34.40,19,7650,1901.0,1129.0,463,1.8200,80100
2,-114.56,33.69,17,720,174.0,333.0,117,1.6509,85700
3,-114.57,33.64,14,1501,337.0,515.0,226,3.1917,73400
4,-114.57,33.57,20,1454,326.0,624.0,262,1.9250,65500


###QUESTION 2 ANSWERS

In [4]:
print(df.shape)

(17020, 9)


The above .shape output shows that we have 17020 Rows and 9 Columns(features)

###QUESTION 3 ANSWERS

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17020 entries, 0 to 17019
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           17020 non-null  float64
 1   latitude            17020 non-null  float64
 2   housing_median_age  17020 non-null  int64  
 3   total_rooms         17020 non-null  int64  
 4   total_bedrooms      16990 non-null  float64
 5   population          16980 non-null  float64
 6   households          17020 non-null  int64  
 7   median_income       16990 non-null  float64
 8   median_house_value  17020 non-null  int64  
dtypes: float64(5), int64(4)
memory usage: 1.2 MB


###QUESTION 4(i) ANSWERS

In [6]:
df.describe()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
count,17020.000000,17020.000000,17020.000000,17020.000000,16990.000000,16980.000000,17020.000000,16990.00000,17020.000000
mean,-119.562066,35.625785,28.590541,2864.361927,539.285756,1429.529505,501.103408,3.88340,205747.585488
std,2.005157,2.137446,12.588782,9385.577088,421.598465,1148.157282,384.400904,1.90864,118642.283445
min,-124.350000,32.540000,1.000000,2.000000,1.000000,3.000000,1.000000,0.49990,-500001.000000
25%,-121.790000,33.930000,18.000000,1462.000000,296.000000,789.000000,282.000000,2.56590,118800.000000
50%,-118.490000,34.250000,29.000000,2127.500000,434.000000,1166.500000,409.000000,3.54425,179800.000000
75%,-118.000000,37.720000,37.000000,3153.000000,648.750000,1721.000000,605.250000,4.76760,264700.000000
max,-114.310000,41.950000,52.000000,379370.000000,6445.000000,35682.000000,6082.000000,15.00010,500001.000000


###QUESTION 4(ii)

One thing I noticed is that the values in the dataset vary a lot from one column to another. Some columns, like longitude and latitude, contain relatively small numbers, while others such as total rooms, population, and median house value have much larger values, ranging from thousands to hundreds of thousands.

###QUESTION 5(i)

In [7]:
df.isnull().sum()

,0
longitude,0
latitude,0
housing_median_age,0
total_rooms,0
total_bedrooms,30
population,40
households,0
median_income,30
median_house_value,0


###QUESTION 5(ii)

(1) total_bedrooms missing values: 30

(2) population missing values: 40

(3) median_income missing values: 30

###QUESTION 6

In [8]:
print(df.duplicated().sum())

20


##SECTION 2: DATA CLEANING

###QUESTION 7(i): Handling missing values

#####TOTAL_BEDROOMS MISSING VALUES

In [9]:
total_bedrooms_median = df['total_bedrooms'].median()
total_bedrooms_median

434.0

In [10]:
df['total_bedrooms'] = df['total_bedrooms'].fillna(total_bedrooms_median)

(1) For the total_bedrooms missing values, I chose to fill it up with median.

(2) Because of Outliers that exist in the column. Median is usually not affected by extreme values

(3) Sanity check to confirm there are no missing values in total_bedrooms

In [11]:
print(df['total_bedrooms'].isnull().sum())

0


#####POPULATION MISSING VALUES

In [12]:
population_median = df['population'].median()
population_median

1166.5

In [13]:
df['population'] = df['population'].fillna(population_median)

(1) For the population missing values, I chose to fill it up with median.

(2) Because of Outliers that exist in the column. Median is usually not affected by extreme values



(3) Sanity check to confirm there are no missing values in population column

In [14]:
print(df['population'].isnull().sum())

0


#####MEDIAN_INCOME MISSING VALUES

In [15]:
median_income_mean = df['median_income'].mean()
print(median_income_mean)

3.883399793996468


In [16]:
df['median_income'] = df['median_income'].fillna(median_income_mean)

(1) For the median_income missing values, I chose to fill it up with mean.

(2) Because there are no Outliers that exist in the column, all the datasets are used here and the precise average is calculated.

(3) Sanity check to confirm there are no missing values in median_income column



In [17]:
print(df['median_income'].isnull().sum())

0


In [18]:
df.isnull().sum()

,0
longitude,0
latitude,0
housing_median_age,0
total_rooms,0
total_bedrooms,0
population,0
households,0
median_income,0
median_house_value,0


###QUESTION 8 ANSWERS

#####Removing duplicates

In [19]:
df.duplicated().sum()

np.int64(20)

In [20]:
df = df.drop_duplicates()

In [21]:
df.shape

(17000, 9)

##SECTION 3: FEATURE SCALING

#####QUESTION 9 ANSWERS

(i) Feature scaling is a technique used in machine learning to bring all numerical features to a similar range of values or scale.

(ii) The problem that will arise without feature scaling is that When one variable has very large numbers (like income) and another has much smaller values (like age), the model may wrongly focus more on the large valued feature, even though it is not actually more important than the other.

##QUESTION 10 ANSWERS

#####question 10, 11(1)

In [29]:
col = 'housing_median_age'
col_min = df[col].min()
col_max = df[col].max()
df[col + '_scaled'] = (df[col] - col_min) / (col_max - col_min)
print(df[col + '_scaled'].min())
print(df[col + '_scaled'].max())


0.0
1.0


#####qusetion 10, 11(2)

In [28]:
col = 'total_rooms'
col_min = df[col].min()
col_max = df[col].max()
df[col + '_scaled'] = (df[col] - col_min) / (col_max - col_min)
print(df[col + '_scaled'].min())
print(df[col + '_scaled'].max())


0.0
1.0


#####question 10, 11(3)

In [27]:
col = 'median_income'
col_min = df[col].min()
col_max = df[col].max()
df[col + '_scaled'] = (df[col] - col_min) / (col_max - col_min)
print(df[col + '_scaled'].min())
print(df[col + '_scaled'].max())


0.0
1.0


#####question 10, 11(4)

In [26]:
col = 'median_house_value'
col_min = df[col].min()
col_max = df[col].max()
df[col + '_scaled'] = (df[col] - col_min) / (col_max - col_min)
print(df[col + '_scaled'].min())
print(df[col + '_scaled'].max())


0.0
1.0


##QUESTION 12 ANSWERS

In [32]:
col = 'median_income'
col_min = df[col].min()
col_max = df[col].max()
df[col + '_scaled'] = (df[col] - col_min) / (col_max - col_min)
print(df[[col, col + '_scaled']].head(10))

   median_income  median_income_scaled
0         1.4936              0.068530
1         1.8200              0.091040
2         1.6509              0.079378
3         3.1917              0.185639
4         1.9250              0.098281
5         3.3438              0.196128
6         2.6768              0.150129
7         1.7083              0.083337
8         2.1782              0.115743
9         2.1908              0.116612


The median_income feature was transformed to values between 0 and 1, making it more consistent with other variables and helping machine learning models learn more effectively.

We can also say that Larger original income values resulted in larger scaled values.